runtime/derive/src/lib.rs

```diff
@@ -65,18 +179,68 @@ fn asc_type_derive_struct(item_struct: ItemStruct) -> TokenStream {
         Fields::Named(fields) => fields.named.iter().map(|field| &field.ty),
         _ => panic!("AscType can only be derived for structs with named fields"),
     };
+    let field_types2 = field_types.clone();
+
+    // if no fields, return immediately
+    let (no_fields_return_bytes, no_fields_return_self) = if field_names.is_empty() {
+        (
+            quote! {
+                #![allow(unreachable_code)]
+                return Ok(Vec::new());
+            },
+            quote! {
+                #![allow(unreachable_code)]
+                return Ok(Self {});
+            },
+        )
+    } else {
+        (quote! {}, quote! {})
+    };
 
     TokenStream::from(quote! {
         impl #impl_generics graph::runtime::AscType for #struct_name #ty_generics #where_clause {
             fn to_asc_bytes(&self) -> Result<Vec<u8>, graph::runtime::DeterministicHostError> {
+                #no_fields_return_bytes
+
                 let in_memory_byte_count = std::mem::size_of::<Self>();
```

Comment:
This gives the Rust `repr(C)` memory size in bytes. Now that we implement a proper padding, the later assertion below (need to expand context in PR to see the line in question):

```
assert_eq!(bytes.len(), in_memory_byte_count, "Alignment mismatch for {}, re-order fields or explicitly add a _padding field", stringify!(#struct_name));
```

Now that we deal with padding ourself, I'm pretty sure the assertion will triggers now in some occasions where padding is different between `repr(C)` and AssemblyScript rules.

We will need to see if Edge&Node here how we are going to ensure this is thoroughly tested because such change if we get it wrong can lead to invisible bugs where the data will be just wrong without any signs of crashes.